rfcs/syntax-default-type-alias-type-parameters.md

```diff
@@ -0,0 +1,93 @@
+# Default type alias type parameters
+
+## Summary
+
+Introduce syntax to provide default type values inside the type alias type parameter list.
+
+## Motivation
+
+Luau has support for type parameters for type aliases and functions.
+In languages with similar features like C++, Rust, Flow and TypeScript, it is possible to specify default values for looser coupling and easier composability, and users with experience in those languages would like to have these design capabilities in Luau.
+
+Here is an example that is coming up frequently during development of GraphQL Luau library:
+```lua
+export type GraphQLFieldResolver<
+    TSource,
+    TContext,
+    TArgs = { [string]: any }
+> = (TSource, TArgs, TContext, GraphQLResolveInfo) -> any
+```
+If we could specify defaults like that, we won't have to write long type names when type alias is used unless specific customization is required.
+Some engineers already skip these extra arguments and use `'any'` to save time, which gives worse typechecking quality.
+
+Without default parameter values it's also harder to refactor the code as each type alias reference that uses 'common' type arguments has to be updated.
+
+While previous example uses a concrete type for default type value, it should also be possible to reference generic types from the same list:
+```lua
+type Eq<T, U = T> = (l: T, r: U) -> boolean
+
+local a: Eq<number> = ...
+local b: Eq<number, string> = ...
+```
+
+Generic functions in Luau also have a type parameter list, but it's not possible to specify type arguments at the call site and because of that, default type parameter values for generic functions are not proposed.
+
+## Design
+
+If a default type parameter value is assigned, following type parameters (on the right) must also have default type parameter values.
+```lua
+type A<T, U = string, V> = ... -- not allowed
+```
+
+Default type parameter values can reference type parameters which were defined earlier (to the left):
+```lua
+type A<T, U = T> = ...-- ok
+
+type A<T, U = V, V = T> = ... -- not allowed
+```
+
+Default type parameter values are also allowed for type packs:
+```lua
+type A<T, U... = ...string>        -- ok, variadic type pack
+type B<T, U... = (number, number)> -- ok, type pack with two elements
```

Comment:
We don't have syntax for this at the moment. IMO this should be a separate proposal.

Reply:
Which part of the syntax exactly?

Reply:
The `(T, T)` syntax.

Reply:
Fixed.
That's from return type annotation tail and is not accessible as a stand-alone syntax. I've removed this line.